cmd/nmi/main.go

```diff
@@ -110,11 +123,30 @@ func main() {
 	// will report "Active" once the iptables rules are set
 	probes.InitAndStart(*httpProbePort, &s.Initialized)
 
-	// Register and expose metrics views
-	if err = metrics.RegisterAndExport(*prometheusPort); err != nil {
-		klog.Fatalf("Could not register and export metrics: %+v", err)
+	mainRoutineDone := make(chan struct{})
+	subRoutineDone := make(chan struct{})
+
+	var redirector server.RedirectorFunc
+	if runtime.GOOS == "windows" {
+		redirector = server.WindowsRedirector(s, subRoutineDone)
+	} else {
+		redirector = server.LinuxRedirector(s, subRoutineDone)
 	}
+
```

Comment:
Looks like you are passing this channel twice to redirectors. First time when creating redirector, second time when you start redirector. Same for the server object.

Better pattern would be to create a channel in redirector constructor and return it back together with redirector. Return read-only channel - this indicates that main function cannot close it only redirector itself can do this. Closing `subroutineDone` channel in main will actually cause panic

As for server object it is better to capture it in constructor, that is returned redirector will already have an instance of `s`
```
       var redirectorDone <-chan struct{}
       if runtime.GOOS == "windows" {
		redirector, redirectorDone = server.WindowsRedirector(s)
	} else {
		redirector, redirectorDone = server.LinuxRedirector(s)
	}


       go redirector(mainDone)
```

Reply:
As discussed, will file a task to track this issue and fix it in next iteration